federation/service/node.go

```diff
@@ -0,0 +1,203 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"strconv"
+	"time"
+
+	"github.com/cortezaproject/corteza-server/federation/types"
+	"github.com/cortezaproject/corteza-server/federation/util"
+	"github.com/cortezaproject/corteza-server/pkg/actionlog"
+	"github.com/cortezaproject/corteza-server/pkg/auth"
+	"github.com/cortezaproject/corteza-server/pkg/id"
+	"github.com/cortezaproject/corteza-server/store"
+	"github.com/cortezaproject/corteza-server/system/service"
+	stypes "github.com/cortezaproject/corteza-server/system/types"
+	"github.com/davecgh/go-spew/spew"
+)
+
+var (
+	// This is temporary for local testing
+	tmpNodeStore = make(types.NodeSet, 0)
+
+	ErrorNodeNotFound   = errors.New("pair: node not found")
+	ErrorInvalidNodeURI = errors.New("pair: invalid node uri provided")
+)
+
+type (
+	node struct {
+		ctx context.Context
+
+		actionlog actionlog.Recorder
+
+		store        store.Storable
+		sysUser      service.UserService
+		tokenEncoder auth.TokenEncoder
+	}
+
+	NodeService interface {
+		With(ctx context.Context) NodeService
+
+		Create(n *types.Node, sharedDomain string) (*types.Node, error)
+		Pair(nodeID uint64) error
+		HandshakeInit(sourceNodeID, nodeIDB uint64, nodeURI, token string) error
+		HandshakeConfirm(nodeID uint64) error
+		HandshakeComplete(nodeID uint64, token string) error
+	}
+)
+
+func Node(ctx context.Context) NodeService {
+	rand.Seed(time.Now().UnixNano())
+	return (&node{
+		store: DefaultNgStore,
+
+		actionlog:    DefaultActionlog,
+		tokenEncoder: auth.DefaultJwtHandler,
+	}).With(ctx)
+}
+
+func (svc node) With(ctx context.Context) NodeService {
+	return &node{
+		ctx: ctx,
+
+		actionlog: svc.actionlog,
+
+		store:        svc.store,
+		sysUser:      service.DefaultUser.With(ctx),
+		tokenEncoder: auth.DefaultJwtHandler,
+	}
+}
+
+// @todo move myDomain to configuration
+func (svc node) Create(n *types.Node, myDomain string) (*types.Node, error) {
+	n.ID = id.Next()
+	n.Status = types.NodeStatusPending
+
+	ott := util.GenerateOTT()
+	i := util.EncodeURI(ott, myDomain, n.ID)
+	if n.NodeURI == "" {
+		n.NodeURI = i
+	}
+
+	// @todo store an initial Node entry
+	tmpNodeStore = append(tmpNodeStore, n)
+
+	return n, nil
+}
+
+func (svc node) Pair(nodeID uint64) error {
+	// @todo store
+	n := tmpNodeStore.FindByID(nodeID)
+	if n == nil {
+		return ErrorNodeNotFound
+	}
+
+	// Handle fenedrated user
+	uHandle := "federation_" + strconv.FormatUint(n.ID, 10)
+	u, _ := svc.sysUser.FindByHandle(uHandle)
+
+	if u == nil {
+		var err error
+
+		// Create a system user to service this node
+		u, err = svc.sysUser.Create(&stypes.User{
+			Email:          strconv.FormatUint(n.ID, 10) + "@federation.corteza",
+			Handle:         uHandle,
+			Kind:           stypes.FederationUser,
```

Comment:
No, let's not add user kinds like this.
IF we only need this to be diferentiated from other users, we'll solve this later with labels.

Reply:
From my point of view, it would make sense to have a new user kind for such a thing. I think that the underlying logic would be cleaner if we solve this with user kinds instead of labels.

- we know what user is a federated user from the kind,
- we can prevent explicit logins for users that are used in federation.

Reply:
I'm not 100% sure yet what we're going to do with service users, bots (if ever) etc... 
For now, this kind "feature" can easily be dropped as we do not use it anywhere.